From 8b5816272945e9f1785ad3171b21341d601b1054 Mon Sep 17 00:00:00 2001
From: Istio Automation <istio-testing-bot@google.com>
Date: Mon, 15 Nov 2021 11:32:44 -0800
Subject: [release-1.12] add downstream_cx_active stats only when
 ExitOnZeroActiveConnections is used (#36089)

* enable downstream_cx_active stats only when ExitOnZeroActiveConnections is enabled

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix tests

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* ignore worker stats

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

Co-authored-by: Rama Chavali <rama.rao@salesforce.com>
---
 pilot/pkg/model/context.go                                 | 3 +++
 pkg/bootstrap/config.go                                    | 6 +++++-
 pkg/bootstrap/instance_test.go                             | 4 ++--
 pkg/bootstrap/testdata/all_golden.json                     | 3 ---
 pkg/bootstrap/testdata/auth_golden.json                    | 3 ---
 pkg/bootstrap/testdata/authsds_golden.json                 | 3 ---
 pkg/bootstrap/testdata/default_golden.json                 | 3 ---
 pkg/bootstrap/testdata/metrics_no_statsd_golden.json       | 3 ---
 pkg/bootstrap/testdata/running_golden.json                 | 3 ---
 pkg/bootstrap/testdata/runningsds_golden.json              | 3 ---
 pkg/bootstrap/testdata/stats_inclusion_golden.json         | 3 ---
 pkg/bootstrap/testdata/tracing_datadog_golden.json         | 3 ---
 pkg/bootstrap/testdata/tracing_lightstep_golden.json       | 3 ---
 pkg/bootstrap/testdata/tracing_opencensusagent_golden.json | 3 ---
 pkg/bootstrap/testdata/tracing_stackdriver_golden.json     | 3 ---
 pkg/bootstrap/testdata/tracing_tls_custom_sni_golden.json  | 3 ---
 pkg/bootstrap/testdata/tracing_tls_golden.json             | 3 ---
 pkg/bootstrap/testdata/tracing_zipkin_golden.json          | 3 ---
 pkg/bootstrap/testdata/xdsproxy_golden.json                | 3 ---
 pkg/envoy/agent.go                                         | 5 +++--
 20 files changed, 13 insertions(+), 53 deletions(-)

diff --git a/pilot/pkg/model/context.go b/pilot/pkg/model/context.go
index 7a2c3fd488..7b8cdfb851 100644
--- a/pilot/pkg/model/context.go
+++ b/pilot/pkg/model/context.go
@@ -621,6 +621,9 @@ type NodeMetadata struct {
 	// Envoy prometheus port redirecting to admin port prometheus endpoint.
 	EnvoyPrometheusPort int `json:"ENVOY_PROMETHEUS_PORT,omitempty"`
 
+	// ExitOnZeroActiveConnections terminates Envoy if there are no active connections if set.
+	ExitOnZeroActiveConnections StringBool `json:"EXIT_ON_ZERO_ACTIVE_CONNECTIONS,omitempty"`
+
 	// Contains a copy of the raw metadata. This is needed to lookup arbitrary values.
 	// If a value is known ahead of time it should be added to the struct rather than reading from here,
 	Raw map[string]interface{} `json:"-"`
diff --git a/pkg/bootstrap/config.go b/pkg/bootstrap/config.go
index ef707bd199..2840286b4a 100644
--- a/pkg/bootstrap/config.go
+++ b/pkg/bootstrap/config.go
@@ -230,12 +230,16 @@ func getStatsOptions(meta *model.BootstrapNodeMetadata) []option.Instance {
 		proxyConfigSuffixes = config.ProxyStatsMatcher.InclusionSuffixes
 		proxyConfigRegexps = config.ProxyStatsMatcher.InclusionRegexps
 	}
+	inclusionSuffixes := rbacEnvoyStatsMatcherInclusionSuffix
+	if meta.ExitOnZeroActiveConnections {
+		inclusionSuffixes = requiredEnvoyStatsMatcherInclusionSuffixes
+	}
 
 	return []option.Instance{
 		option.EnvoyStatsMatcherInclusionPrefix(parseOption(prefixAnno,
 			requiredEnvoyStatsMatcherInclusionPrefixes, proxyConfigPrefixes)),
 		option.EnvoyStatsMatcherInclusionSuffix(parseOption(suffixAnno,
-			requiredEnvoyStatsMatcherInclusionSuffixes, proxyConfigSuffixes)),
+			inclusionSuffixes, proxyConfigSuffixes)),
 		option.EnvoyStatsMatcherInclusionRegexp(parseOption(RegexAnno, "", proxyConfigRegexps)),
 		option.EnvoyExtraStatTags(extraStatTags),
 	}
diff --git a/pkg/bootstrap/instance_test.go b/pkg/bootstrap/instance_test.go
index ebf3306ef1..47fc469d0c 100644
--- a/pkg/bootstrap/instance_test.go
+++ b/pkg/bootstrap/instance_test.go
@@ -472,9 +472,9 @@ func checkStatsMatcher(t *testing.T, got, want *bootstrap.Bootstrap, stats stats
 		stats.prefixes = v2Prefixes + stats.prefixes + "," + requiredEnvoyStatsMatcherInclusionPrefixes + v2Suffix
 	}
 	if stats.suffixes == "" {
-		stats.suffixes = requiredEnvoyStatsMatcherInclusionSuffixes
+		stats.suffixes = rbacEnvoyStatsMatcherInclusionSuffix
 	} else {
-		stats.suffixes += "," + requiredEnvoyStatsMatcherInclusionSuffixes
+		stats.suffixes += "," + rbacEnvoyStatsMatcherInclusionSuffix
 	}
 
 	if err := gsm.Validate(); err != nil {
diff --git a/pkg/bootstrap/testdata/all_golden.json b/pkg/bootstrap/testdata/all_golden.json
index 145cff097c..9f1fbbd48d 100644
--- a/pkg/bootstrap/testdata/all_golden.json
+++ b/pkg/bootstrap/testdata/all_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/auth_golden.json b/pkg/bootstrap/testdata/auth_golden.json
index dbf18ba0c8..53bf542ac5 100644
--- a/pkg/bootstrap/testdata/auth_golden.json
+++ b/pkg/bootstrap/testdata/auth_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/authsds_golden.json b/pkg/bootstrap/testdata/authsds_golden.json
index 8594ddde33..390e0a2cc0 100644
--- a/pkg/bootstrap/testdata/authsds_golden.json
+++ b/pkg/bootstrap/testdata/authsds_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/default_golden.json b/pkg/bootstrap/testdata/default_golden.json
index f3b26c2069..eb63627c75 100644
--- a/pkg/bootstrap/testdata/default_golden.json
+++ b/pkg/bootstrap/testdata/default_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/metrics_no_statsd_golden.json b/pkg/bootstrap/testdata/metrics_no_statsd_golden.json
index c30e4090ad..af02142e39 100644
--- a/pkg/bootstrap/testdata/metrics_no_statsd_golden.json
+++ b/pkg/bootstrap/testdata/metrics_no_statsd_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/running_golden.json b/pkg/bootstrap/testdata/running_golden.json
index 95fb22f5eb..23d0a3e116 100644
--- a/pkg/bootstrap/testdata/running_golden.json
+++ b/pkg/bootstrap/testdata/running_golden.json
@@ -248,9 +248,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/runningsds_golden.json b/pkg/bootstrap/testdata/runningsds_golden.json
index e2381e8dfa..067ae1308a 100644
--- a/pkg/bootstrap/testdata/runningsds_golden.json
+++ b/pkg/bootstrap/testdata/runningsds_golden.json
@@ -248,9 +248,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/stats_inclusion_golden.json b/pkg/bootstrap/testdata/stats_inclusion_golden.json
index f7b2c5d4f4..408aa0828e 100644
--- a/pkg/bootstrap/testdata/stats_inclusion_golden.json
+++ b/pkg/bootstrap/testdata/stats_inclusion_golden.json
@@ -333,9 +333,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "safe_regex": {"google_re2":{}, "regex":"http.[0-9]*\\.[0-9]*\\.[0-9]*\\.[0-9]*_8080.downstream_rq_time"}
           },
           {
diff --git a/pkg/bootstrap/testdata/tracing_datadog_golden.json b/pkg/bootstrap/testdata/tracing_datadog_golden.json
index 3d0dc39b3c..690494c295 100644
--- a/pkg/bootstrap/testdata/tracing_datadog_golden.json
+++ b/pkg/bootstrap/testdata/tracing_datadog_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/tracing_lightstep_golden.json b/pkg/bootstrap/testdata/tracing_lightstep_golden.json
index 097d28d807..e040b8dcf7 100644
--- a/pkg/bootstrap/testdata/tracing_lightstep_golden.json
+++ b/pkg/bootstrap/testdata/tracing_lightstep_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/tracing_opencensusagent_golden.json b/pkg/bootstrap/testdata/tracing_opencensusagent_golden.json
index bb5fa8460b..4f981d77e5 100644
--- a/pkg/bootstrap/testdata/tracing_opencensusagent_golden.json
+++ b/pkg/bootstrap/testdata/tracing_opencensusagent_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/tracing_stackdriver_golden.json b/pkg/bootstrap/testdata/tracing_stackdriver_golden.json
index 277cc0e6ba..943369ed7c 100644
--- a/pkg/bootstrap/testdata/tracing_stackdriver_golden.json
+++ b/pkg/bootstrap/testdata/tracing_stackdriver_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/tracing_tls_custom_sni_golden.json b/pkg/bootstrap/testdata/tracing_tls_custom_sni_golden.json
index 9dd8c2ff64..6d6e8e5108 100644
--- a/pkg/bootstrap/testdata/tracing_tls_custom_sni_golden.json
+++ b/pkg/bootstrap/testdata/tracing_tls_custom_sni_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/tracing_tls_golden.json b/pkg/bootstrap/testdata/tracing_tls_golden.json
index 815e8fc8d0..919b1e6bb9 100644
--- a/pkg/bootstrap/testdata/tracing_tls_golden.json
+++ b/pkg/bootstrap/testdata/tracing_tls_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/tracing_zipkin_golden.json b/pkg/bootstrap/testdata/tracing_zipkin_golden.json
index 48fa00e217..e4ca15d1d0 100644
--- a/pkg/bootstrap/testdata/tracing_zipkin_golden.json
+++ b/pkg/bootstrap/testdata/tracing_zipkin_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/bootstrap/testdata/xdsproxy_golden.json b/pkg/bootstrap/testdata/xdsproxy_golden.json
index 27c8fdd15a..1a448076ff 100644
--- a/pkg/bootstrap/testdata/xdsproxy_golden.json
+++ b/pkg/bootstrap/testdata/xdsproxy_golden.json
@@ -243,9 +243,6 @@
           "suffix": "shadow_denied"
           },
           {
-          "suffix": "downstream_cx_active"
-          },
-          {
           "prefix": "component"
           }
         ]
diff --git a/pkg/envoy/agent.go b/pkg/envoy/agent.go
index f2196e266c..8edfbb09fd 100644
--- a/pkg/envoy/agent.go
+++ b/pkg/envoy/agent.go
@@ -183,8 +183,9 @@ func (a *Agent) activeProxyConnections() int {
 			continue
 		}
 		// downstream_cx_active is accounted under "http." and "listener." for http listeners.
-		// Only consider listener stats.
-		if !strings.HasPrefix(parts[0], "listener.") {
+		// Only consider listener stats. Listener stats also will have per worker stats, we can
+		// ignore them.
+		if !strings.HasPrefix(parts[0], "listener.") || strings.Contains(parts[0], "worker_") {
 			continue
 		}
 		// If the stat is for a known Istio listener skip it.
-- 
2.35.3

